### PR TITLE
Additional Workshed Support

### PR DESCRIPTION
**Pull request by Xyghra**
should allow support for CMC + Trainset in addition to asdon. Right now it simply prioritizes asdon > cmc > trainset but I have no calculations to support this choice, It has very little testing since swapping worksheds is a once/day deal and I don't own an asdon

### CODE REVIEW

File: src/quests/prologue/tasks/asdon.ts [hunk 1]
```diff
@@ -3,10 +3,14 @@ import { getFuel, use } from "kolmafia";
...
 import { ASDON } from "../../../prefs/properties";
+import { CMCInstalled } from "./cmc";
 
 export const installAsdon: Task = {
   name: "Install Asdon",
-  completed: () => get("_workshedItemUsed") || AsdonInstalled(),
+  completed: () =>
+    get("_workshedItemUsed") ||
+    AsdonInstalled() ||
+    (CMCInstalled() && get(`_coldMedicineConsults`) < 5),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/asdon.ts]: Let's keep CMC out of this file.

Can do something like
```
ready: () => get(ASDON)
```

File: src/quests/prologue/tasks/cmc.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import { Task } from "grimoire-kolmafia";
+import { getWorkshed, myAdventures, runChoice, totalTurnsPlayed, use, visitUrl } from "kolmafia";
+import { $item, get, have } from "libram";
+
+import { CMC } from "../../../prefs/properties";
+
+export function CMCInstalled(): boolean {
+  if (!get(CMC)) {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: This is misleading logic in a function named `CMCInstalled`

File: src/quests/prologue/tasks/cmc.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
+import { Task } from "grimoire-kolmafia";
+import { getWorkshed, myAdventures, runChoice, totalTurnsPlayed, use, visitUrl } from "kolmafia";
+import { $item, get, have } from "libram";
+
+import { CMC } from "../../../prefs/properties";
+
+export function CMCInstalled(): boolean {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: This function doesn't seem to be checking if the CMC is actually installed. Please reconsider the function name, or move around the function logic to better reflect this method signature.

File: src/quests/prologue/tasks/cmc.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  if (getWorkshed() === $item`cold medicine cabinet`) {
+    return true;
+  }
+  return !have($item`cold medicine cabinet`);
+}
+
+export const installCMC: Task = {
+  name: `Install CMC`,
+  completed: () => get(`_workshedItemUsed`) || myAdventures() > 81 || CMCInstalled(),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: Can we move the `myAdventures` logic to a `ready` clause?

File: src/quests/prologue/tasks/cmc.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  return [`Breathitin`, `Extrovermectin`].some((pill) => workshedStatus.includes(pill));
+}
+
+export const grabCMCPill: Task = {
+  name: `Grab CMC Pill`,
+  completed: () =>
+    get(`_coldMedicineConsults`) >= 5 ||
+    totalTurnsPlayed() < get(`_nextColdMedicineConsult`) ||
+    !CMCInstalled() ||
+    !CMCPillReady(),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: Let's move this to the ready clause.

File: src/quests/prologue/tasks/train.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+  TrainSet.Station.GAIN_MEAT,
+  TrainSet.Station.VIEWING_PLATFORM,
+];
+
+export const installTrain: Task = {
+  name: `Install Train`,
+  completed: () =>
+    get(`_workshedItemUsed`) ||
+    trainInstalled() ||
+    (CMCInstalled() && get(`_coldMedicineConsults`) < 5),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/train.ts]: Let's keep CMC out of this method. With 3 worksheds supported, I think it makes sense to refactor the `useAsdon` property to a `worshed` property, accepting `train,cmc,asdon`.

File: src/quests/prologue/tasks/train.ts [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+export const reconfigureTrain: Task = {
+  name: `Reconfigure Trainset`,
+  ready: () =>
+    [
+      TrainSet.Station.GRAIN_SILO,
+      TrainSet.Station.TOWER_FIZZY,
+      TrainSet.Station.GAIN_MEAT,
+      TrainSet.Station.VIEWING_PLATFORM,
+    ].includes(TrainSet.next()),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/train.ts]: I'm not sure I understand why this is the ready condition.

**Review by tigerinanenet — CHANGES_REQUESTED** [review]

**Comment by Xyghra** [comment]
I think this resolves most of the requested changes. The first one, I'm just genuinely clueless on how to check and make sure I'm not overwriting CMC early without including any CMC related functions in the file. I did the best I could with what I know how to do in the end, apologies if it's not to your liking.

File: src/quests/prologue/tasks/asdon.ts [hunk 8]
```diff
@@ -2,20 +2,33 @@ import { Task } from "grimoire-kolmafia";
...
-import { ASDON } from "../../../prefs/properties";
+import { WORKSHED } from "../../../prefs/properties";
+import { CMCInProgress } from "./cmc";
+
+function shouldInstallAsdon(): boolean {
+  if (get("_workshedItemUsed")) return false;
+
+  const desiredWorksheds = get(WORKSHED).split(`,`);
+  //Check if user wants to install asdon at all
+  if (desiredWorksheds.includes(`asdon`)) {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/asdon.ts]: Nested if statements are really tricky to understand. We can use the guard pattern here. When cleaned up, the code may look something like this:
```
if (!desiredWorksheds.includes(`asdon`)) {
    return false;
}
if (have($item`Asdon Martin keyfob`)) {
    return false
}
return CMCInProgress();

File: src/quests/prologue/tasks/asdon.ts [hunk 9]
```diff
@@ -2,20 +2,33 @@ import { Task } from "grimoire-kolmafia";
...
-      return;
-    }
-    if (!get(ASDON)) {
-      return;
-    }
-    use($item`Asdon Martin keyfob`);
-  },
+  ready: () => shouldInstallAsdon(),
+  completed: () => AsdonMartin.installed(),
+  do: () => use($item`Asdon Martin keyfob`),
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/asdon.ts:22]: nice cleanup! :+1:

File: src/quests/prologue/tasks/cmc.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import { getWorkshed, myAdventures, runChoice, totalTurnsPlayed, use, visitUrl } from "kolmafia";
+import { $item, get, have } from "libram";
+
+import { WORKSHED } from "../../../prefs/properties";
+
+function shouldInstallCMC(): boolean {
+  if (get("_workshedItemUsed")) return false;
+
+  //Check if user wants to use CMC
+  if (get(WORKSHED).split(`,`).includes(`cmc`)) {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: We can use the guard pattern here as well.

File: src/quests/prologue/tasks/cmc.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    //We need to have it to be able to install it
+    if (have($item`cold medicine cabinet`)) {
+      //Install during the last 80 advs to grab all 5 pills
+      if (myAdventures() <= 80) return true;
+    }
+  }
+  return false;
+}
+function CMCInstalled(): boolean {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts]: It's a little weird, but the convention is to have even abbreviations be lowercase if they start a signature/method name
```
cmcInstalled
```
would be the preferred name.

File: src/quests/prologue/tasks/cmc.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+export const grabCMCPill: Task = {
+  name: `Grab CMC Pill`,
+  ready: () => CMCPillReady() && CMCInstalled(),
+  completed: () =>
+    get(`_coldMedicineConsults`) >= 5 || totalTurnsPlayed() < get(`_nextColdMedicineConsult`),
+  do: (): void => {
+    visitUrl("campground.php?action=workshed");
+    runChoice(5);
+  },
+};
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/cmc.ts:43]: nice!

File: src/quests/prologue/tasks/train.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  TrainSet.Station.CANDY_FACTORY,
+  TrainSet.Station.GRAIN_SILO,
+  TrainSet.Station.ORE_HOPPER,
+  TrainSet.Station.TRACKSIDE_DINER,
+  TrainSet.Station.TOWER_FIZZY,
+  TrainSet.Station.VIEWING_PLATFORM,
+  TrainSet.Station.PRAWN_SILO,
+];
+
+function shouldInstallTrain(): boolean {
```
Comment (tigerinanenet) [inline src/quests/prologue/tasks/train.ts:26]: Lets use the guard pattern here too.

**Review by tigerinanenet — CHANGES_REQUESTED** [review]

**Comment by Xyghra** [comment]
👍

**Pull request closed without merging**